## Task (a): Handling Missing Values

In this step, I checked for missing values in the dataset and decided how to handle them. I first explored the dataset to understand which columns have missing values and how many.

In [16]:
import pandas as pd

# Load dataset
df = pd.read_csv('../data/train.csv')
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5847 entries, 0 to 5846
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         5847 non-null   int64  
 1   Name               5847 non-null   object 
 2   Location           5847 non-null   object 
 3   Year               5847 non-null   int64  
 4   Kilometers_Driven  5847 non-null   int64  
 5   Fuel_Type          5847 non-null   object 
 6   Transmission       5847 non-null   object 
 7   Owner_Type         5847 non-null   object 
 8   Mileage            5845 non-null   object 
 9   Engine             5811 non-null   object 
 10  Power              5811 non-null   object 
 11  Seats              5809 non-null   float64
 12  New_Price          815 non-null    object 
 13  Price              5847 non-null   float64
dtypes: float64(2), int64(3), object(9)
memory usage: 639.6+ KB


In [18]:
df.isnull().sum()

Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64

From this output, I saw that:
- New_Price has more than 5000 missing values out of 5847 rows (>85%), so I dropped it.
- Mileage, Engine, Power, and Seats have some missing values, but not too many.

In [19]:
df.drop(columns=['New_Price'], inplace=True)
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  Price  
0  126.2 bhp    5.0  12.50  
1   88.7 bhp    5.0   4.50  
2  88.76 bhp    7.0   6.00  
3  140.8 bhp    5.0  17.74  
4   63.1 bhp    5.0   3.50

Before filling missing values in Mileage, Engine, and Power, I removed units from those columns to make them numeric.

In [20]:
df['Mileage'] = df['Mileage'].str.extract(r'(\d+\.?\d*)').astype(float)
df['Engine'] = df['Engine'].str.extract(r'(\d+\.?\d*)').astype(float)
df['Power'] = df['Power'].str.extract(r'(\d+\.?\d*)').astype(float)

In [21]:
df['Mileage'].fillna(df['Mileage'].mode()[0], inplace=True)
df['Engine'].fillna(df['Engine'].mode()[0], inplace=True)
df['Power'].fillna(df['Power'].mode()[0], inplace=True)
df['Seats'].fillna(df['Seats'].mode()[0], inplace=True)

df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type  Mileage  Engine  \
0              41000    Diesel       Manual      First    19.67  1582.0   
1              46000    Petrol       Manual      First    13.00  1199.0   
2              87000    Diesel       Manual      First    20.77  1248.0   
3              40670    Diesel    Automatic     Second    15.20  1968.0   
4              86999    Diesel       Manual      First    23.08  1461.0   

    Power  Seats  Price  
0  126.20    5.0  12.50  
1   88.70    5.0   4.50  
2   88.76    7.0   6.00  
3  140.80    5.0  17.74  
4   63.10    5.0   3.50

I used **mode** for these columns because values like engine CC and seat count are often standard and repeated (like 1199 CC or 5 seats). Using mode keeps the data realistic compared to using average values like mean or median.

## Task (b): Removing Units

In task (a), I already removed the units from Mileage, Engine, and Power. I’m just printing some values here to confirm the result.

In [22]:
df[['Mileage', 'Engine', 'Power']].head()

Mileage  Engine   Power
0    19.67  1582.0  126.20
1    13.00  1199.0   88.70
2    20.77  1248.0   88.76
3    15.20  1968.0  140.80
4    23.08  1461.0   63.10

## Task (c): One-Hot Encoding

The columns Fuel_Type and Transmission are text values. I used one-hot encoding to convert them into numeric format for analysis.

In [23]:
df = pd.get_dummies(df, columns=['Fuel_Type', 'Transmission'])
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0              True               False             False   
1             False               False              True   
2              True               False             False   
3              True               False             False   
4              True               False             False   

   Transmission_Automatic  Transmission_Manual  
0                   False                 True  
1                   False                 True  
2                   False                 True  
3                    True                False  
4                   False                 True

## Task (d): Adding a New Feature

I created a new column called Car_Age which is the age of the car by subtracting its manufacturing year from 2025.

In [24]:
df['Car_Age'] = 2025 - df['Year']
df[['Year', 'Car_Age']].head()

Year  Car_Age
0  2015       10
1  2011       14
2  2012       13
3  2013       12
4  2013       12

## Task (e): DataFrame Operations

In this step, I:
- Renamed Kilometers_Driven to KMs_Driven
- Selected a few important columns
- Filtered cars that were driven more than 100000 km
- Sorted the data by price
- Grouped by location to get average price and number of cars

In [25]:
df.rename(columns={'Kilometers_Driven': 'KMs_Driven'}, inplace=True)

In [26]:
selected = df[['Name', 'Location', 'Year', 'KMs_Driven', 'Fuel_Type_Diesel', 'Fuel_Type_Petrol', 'Price']]
selected.head()

Name    Location  Year  KMs_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015       41000   
1                      Honda Jazz V     Chennai  2011       46000   
2                 Maruti Ertiga VDI     Chennai  2012       87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013       40670   
4            Nissan Micra Diesel XV      Jaipur  2013       86999   

   Fuel_Type_Diesel  Fuel_Type_Petrol  Price  
0              True             False  12.50  
1             False              True   4.50  
2              True             False   6.00  
3              True             False  17.74  
4              True             False   3.50

In [27]:
high_mileage = selected[selected['KMs_Driven'] > 100000]
high_mileage.head()

Name   Location  Year  KMs_Driven  \
13          Mitsubishi Pajero Sport 4X4      Delhi  2014      110000   
27  Toyota Innova 2.5 V Diesel 7-seater     Mumbai  2007      262000   
56                Nissan X-Trail SLX AT  Hyderabad  2010      121812   
62                   Tata Indica V2 eLS    Chennai  2016      178000   
74                 Toyota Innova 2.0 G1    Chennai  2006      230000   

    Fuel_Type_Diesel  Fuel_Type_Petrol  Price  
13              True             False  15.00  
27              True             False   4.00  
56              True             False   7.75  
62              True             False   2.50  
74             False              True   4.50

In [28]:
sorted_df = df.sort_values(by='Price', ascending=False)
sorted_df[['Name', 'Price']].head()

Name   Price
3952  Land Rover Range Rover 3.0 Diesel LWB Vogue  160.00
5620                   Lamborghini Gallardo Coupe  120.00
5752                       Jaguar F Type 5.0 V8 S  100.00
1457              Land Rover Range Rover Sport SE   97.07
1917                           BMW 7 Series 740Li   93.67

In [29]:
summary = df.groupby('Location')['Price'].agg(['mean', 'count']).reset_index()
summary.columns = ['Location', 'Avg_Price', 'Total_Cars']
summary.to_csv('../output/summary_by_location.csv', index=False)
summary.head()

Location  Avg_Price  Total_Cars
0   Ahmedabad   8.567248         218
1   Bangalore  13.482670         352
2     Chennai   7.958340         476
3  Coimbatore  15.160206         631
4       Delhi   9.881944         540

## Conclusion

Each task (a to e) was handled step-by-step in this notebook. I cleaned the data, filled missing values, created a new column, converted text to numbers, and performed basic analysis. The final grouped output is saved as summary_by_location.csv.